src/main/scala/strawman/collection/mutable/Map.scala

```diff
@@ -65,6 +65,18 @@ trait MapOps[K, V, +CC[X, Y] <: MapOps[X, Y, CC, _], +C <: MapOps[K, V, CC, C]]
       case None => val d = op; this(key) = d; d
     }
 
+  /** Removes a key from this map, returning the value associated previously
+    *  with that key as an option.
+    *  @param    key the key to be removed
+    *  @return   an option value containing the value associated previously with `key`,
+    *            or `None` if `key` was not defined in the map before.
+    */
+  def remove(key: K): Option[V] = {
+    val r = get(key)
+    this -= key
```

Comment:
This is not a very efficient way to do this.  You should at least check whether `r` is empty before subtracting `key`, and better is to do it in a single operation.  (That's kind of the point of this method--it's supposed to be the efficient low-level operation in Java maps.)

Reply:
Do you have an idea of how to implement it generically? I think we want to override this default implementation in each specific Map implementation.